src/hardware/uni-t-ut8802e/protocol.c

```diff
@@ -0,0 +1,430 @@
+            break;
+
+        case MODE_DIODE:
+            info->meas_head.mode = SR_MQ_VOLTAGE;
+            info->meas_head.mqflag = SR_MQFLAG_DIODE | SR_MQFLAG_DC;
+            info->meas_data.main_unit = SR_MQ_VOLTAGE;
+            break;
+
+        case MODE_CAPACITANCE_nF:
+            info->meas_head.mode = SR_MQ_CAPACITANCE;
+            info->meas_data.main_unit = SR_UNIT_FARAD;
+            info->meas_data.main_prec = range_farad[0];
+            break;
+        case MODE_CAPACITANCE_uF:
+            info->meas_head.mode = SR_MQ_CAPACITANCE;
+            info->meas_data.main_unit = SR_UNIT_FARAD;
+            info->meas_data.main_prec = range_farad[1];
+            break;
+        case MODE_CAPACITANCE_mF:
+            info->meas_head.mode = SR_MQ_CAPACITANCE;
+            info->meas_data.main_unit = SR_UNIT_FARAD;
+            info->meas_data.main_prec = range_farad[2];
+            break;
+
+        case MODE_TRIODE_HFE:
+            info->meas_head.mode = SR_MQ_GAIN;
+            info->meas_data.main_unit = SR_UNIT_UNITLESS;
+            break;
+
+        case MODE_THYRISTOR_SCR:
+            info->meas_head.mode = SR_MQ_VOLTAGE;
+            info->meas_data.main_unit = SR_UNIT_VOLT;
+            break;
+
+        case MODE_FREQ_Hz:
+            info->meas_head.mode = SR_MQ_FREQUENCY;
+            info->meas_data.main_unit = SR_UNIT_HERTZ;
+            info->meas_data.main_prec = range_hz[0];
+            break;
+        case MODE_FREQ_KHz:
+            info->meas_head.mode = SR_MQ_FREQUENCY;
+            info->meas_data.main_unit = SR_UNIT_HERTZ;
+            info->meas_data.main_prec = range_hz[1];
+            break;
+        case MODE_FREQ_MHz:
+            info->meas_head.mode = SR_MQ_FREQUENCY;
+            info->meas_data.main_unit = SR_UNIT_HERTZ;
+            info->meas_data.main_prec = range_hz[2];
+            break;
+
+        case MODE_DUTY:
+            info->meas_head.mode = SR_MQ_DUTY_CYCLE;
+            info->meas_data.main_unit = SR_UNIT_PERCENTAGE;
+            break;
+
+        default:
+            sr_spew("Unkwon functionality");
+            return SR_ERR_DATA;
+    }
+
+    ret = ut8802e_feedbuff_initialization(&feedbuff);
+    if (ret != SR_OK)
+        return SR_ERR_DATA;
+
+    g_slist_free(feedbuff.analog.meaning->channels);
+    feedbuff.analog.meaning->channels = g_slist_append(
+        NULL,
+        g_slist_nth_data(sdi->channels, ut8802e_CH_MAIN));
+
+    feedbuff.analog.meaning->mqflags = info->meas_head.mqflag;
+    feedbuff.analog.meaning->mq = info->meas_head.mode;
+    feedbuff.analog.meaning->unit = info->meas_data.main_unit;
+
+    info->meas_data.main_value = new_value * pow(10, info->meas_data.main_prec);
+
+    feedbuff.analog.encoding->digits = -info->meas_data.main_prec + 4;
+    feedbuff.analog.spec->spec_digits = -info->meas_data.main_prec + 4;
+
+    feedbuff.main_value = info->meas_data.main_value;
+	if (sdi->status != SR_ST_ACTIVE)
```

Comment:
I don't see any reason for this indentation.

Reply:
Fixed